design/multi-checkout.md

```diff
@@ -0,0 +1,201 @@
+
+`System.DefaultWorkingDirectory` points to the same place as `Pipeline.Workspace` in multi-checkout mode (which is different than single-checkout mode).
+We believe this preserves the intended semantics of these variables, keeping as many scripts and tasks working as possible.
+
+## Resource specification
+
+The `resources` block already allows a `repositories` list.
+Today, those repositories hold only YAML templates, so we can't blindly check them out automatically.
+That's why an explicit `checkout` step is also required.
+
+However, in multi-checkout scenarios, this leads to unnecessary verbosity.
+First you have to mention the repository up top, give it a name, and then you check it out by name.
+To combat this, we'll add an inline repository syntax.
+Inline syntax will cover the `type`, `name`, and `ref` of a repository resource.
+(If other changes are needed - such as `endpoint` - users must fall back to explicit resource syntax.)
+
+## Supported repo types
+
+Supported repo types will match what YAML supports:
+- Azure Repos
+- GitHub
+- External Git (but not for the triggering repo)
+
+If and when other providers are supported for YAML, they will be supported by this feature as well.
+
+## Syntax examples
+
+### Basic syntax with inline resource
+
+The most basic, and probably most common form of multi-checkout:
+```yaml
+steps:
+- checkout: self
+- checkout: git://MyProject/MyToolsRepo
+```
+
+This grabs a tools repository from Azure Repos and checks out its default branch, typically `master`.
+Assuming this is in a repository called "MySourceRepo", the resulting directory structure on a hosted agent would be:
+
+```
+_w/
+   1/
+     MySourceRepo/
+         (files from MySourceRepo)
+     MyToolsRepo/
+         (files from MyToolsRepo)
+```
+
+### Basic syntax, GitHub source provider
+
+The same syntax works for GitHub:
+
+```yaml
+steps:
+- checkout: self
+- checkout: github://MyGitHubOrg/MyToolsRepo
+```
+
+And this yields the same directory structure.
+
+### Control directory structure
+
+You can tell the `checkout` step where to put your repos.
+
+```yaml
+steps:
+- checkout: self
+  path: src
+- checkout: github://MyGitHubOrg/MyToolsRepo
+  path: toolzdir
+```
+
+```
+_w/
+   1/
+     src/
+         (files from MySourceRepo)
+     toolzdir/
+         (files from MyToolsRepo)
+```
+
+### Inline ref syntax
+
+```yaml
+steps:
+- checkout: self
+- checkout: git://MyProject/MyToolsRepo@features/mytools
+```
+
+By appending `@<ref>`, the agent can be instructed to check out a different ref.
+In this case, it's assumed to be a branch called `features/mytools`.
+Branches can also be prefixed with `refs/heads/` to make it unambiguous that it's a branch.
+
+Other valid refs (and ref-like things):
+- `refs/tags/<tagname>`
+- commit ID
+
+### Non-inline syntax
+
+If you need an endpoint or other extended resource field, you must use explicit resource syntax.
+
+```yaml
+resources:
+  repositories:
```

Comment:
Instead of having explicit checkout steps inside jobs, another option is to automatically clone all the repositories declared under the repositories keyword. That way, we'll be able to highlight this repositories as first class Git artifacts at the pipeline run details level. Also all agent jobs within that pipeline need not have multiple checkout steps in each of them

Reply:
Two problems I can see:
1. Existing pipelines use repo resources only for templates. It would be a surprising change to have those all suddenly start getting cloned.
2. In a multi-job pipeline, it's likely that you don't need every repo for every job. Thus, you end up needing to say which repos you _don't_ want instead of the ones you do.

Reply:
(By the way, I wish we could do this :grin: and if we ever do a "YAMLv2", we will. But for now, we can't.)